keps/sig-node/4540-strict-cpu-reservation/README.md

```diff
@@ -0,0 +1,465 @@
+
+### Risks and Mitigations
+
+The feature is isolated to a specific policy option within the `CPUManager`, and is protected both by the option itself, as well as the `CPUManagerPolicyAlphaOptions` feature gate (which is disabled by default).
+
+Kube-scheduler schedules pods on node allocatable which is total - reserved cores. It is at the node level, burstable and best-effort pods are allowed to run on the reserved cores. The `strict-cpu-reservation` feature removes this discrepancy.
+
+However, kubelet has been requiring non-zero shared pool when the static policy is enabled. Kube-scheduler knows a portion of the node allocatable is not available for exclusive allocation. This not-for-exclusive portion has been conveniently the reserved cores.
+
+To maintain backward compatibility, we introduce `numMinSharedCPUs` in KubeletConfiguration as the minimum number of CPU cores not available for exclusive allocation and expose it to Kube-scheduler.
+
+![MinSharedCPUs](./strict-cpu-allocation.png)
+
+When `strict-cpu-reservation` is disabled:
+```console
+Total CPU cores: 64
+ReservedSystemCPUs: 6
+defaultCPUSet = Reserved (6) + 58 (available for exclusive allocation)
+```
+
+When `strict-cpu-reservation` is enabled:
+```console
+Total CPU cores: 64
+ReservedSystemCPUs: 6
+MinSharedCPUs: 4
+defaultCPUSet = MinSharedCPUs (4) + 54 (available for exclusive allocation)
+```
+
+## Design Details
+
+In Kubelet, when `strict-cpu-reservation` is enabled as a policy option, we remove the reserved cores from the shared pool at the stage of calculation DefaultCPUSet and remove the `MinSharedCPUs` from the list of available cores for exclusive allocation.
+
+Feature impact can be illustrated as following:
+
+With the following Kubelet configuration:
+
+```yaml
+kind: KubeletConfiguration
+apiVersion: kubelet.config.k8s.io/v1beta1
+featureGates:
+  ...
+  CPUManagerPolicyOptions: true
+  CPUManagerPolicyAlphaOptions: true
+cpuManagerPolicy: static
+cpuManagerPolicyOptions:
+  strict-cpu-reservation: "true"
+reservedSystemCPUs: "0,32,1,33,16,48"
+numMinSharedCPUs: 4
+...
+```
+
+When `strict-cpu-reservation` is disabled:
+```console
+# cat /var/lib/kubelet/cpu\_manager\_state
+{"policyName":"static","defaultCpuSet":"0-79","checksum":1241370203}
+```
+
+When `strict-cpu-reservation` is enabled:
+```console
+# cat /var/lib/kubelet/cpu\_manager\_state
+{"policyName":"static","defaultCpuSet":"2-3,6-15,17-31,34-35,38-47,49-63","entries":{"6dda0e2e-ac6c-4f9d-8cff-201e18aebb2f":{"busybox":"36"},"6ed17b91-0eb0-4a0c-8841-3565fd2e45dc":{"busybox":"37"},"716e4823-ca49-4be7-a429-58d6f3c29d96":{"busybox":"4"},"93ad7328-0dd5-4f9c-ac51-eb46debb16a7":{"busybox":"5"}},"checksum":2153668133}
+```
+
+In Node API, we add `exclusive-cpu` in Node Allocatable for Kube-scheduler to consume.
```

Comment:
will we add a new resource conditionally? Adding a new resource has implications, and this is particularly tricky because it has to be kept in sync with `cpu`, being the two different views of the same resource.

Now: I know this is a pretty straightforward approach to solve the problem presented in this KEP, yet the implications are very deep. We need at least sig-arch ack if we want to go this direction.

Reply:
Your concern is understood, I leave this comment open for getting sig-arch review. My take is that a good design starts with correctness, the current issue is too fundamental, "everybody" is surprised reserved cpus are not actually reserved. We can simply bug-fix this with trivial change in CPUManager itself, but as we agreed, if we want to have this MinSharedCPUs, kube-sceduler has to know it, we want to align scheduler-node view, we cannot want to further skew it.